src/apprt/gtk-ng/class/application.zig

```diff
@@ -0,0 +1,446 @@
+        var err_: ?*glib.Error = null;
+        if (self.as(gio.Application).register(
+            null,
+            &err_,
+        ) == 0) {
+            if (err_) |err| {
+                defer err.free();
+                log.warn(
+                    "error registering application: {s}",
+                    .{err.f_message orelse "(unknown)"},
+                );
+            }
+
+            return error.ApplicationRegisterFailed;
+        }
+        assert(err_ == null);
+
+        // This just calls the `activate` signal but its part of the normal startup
+        // routine so we just call it, but only if the config allows it (this allows
+        // for launching Ghostty in the "background" without immediately opening
+        // a window). An initial window will not be immediately created if we were
+        // launched by D-Bus activation or systemd.  D-Bus activation will send it's
+        // own `activate` or `new-window` signal later.
+        //
+        // https://gitlab.gnome.org/GNOME/glib/-/blob/bd2ccc2f69ecfd78ca3f34ab59e42e2b462bad65/gio/gapplication.c#L2302
+        const priv = self.private();
+        const config = priv.config;
+        if (config.@"initial-window") switch (config.@"launched-from".?) {
+            .desktop, .cli => self.as(gio.Application).activate(),
+            .dbus, .systemd => {},
+        };
+
+        // If we are NOT the primary instance, then we never want to run.
+        // This means that another instance of the GTK app is running and
+        // our "activate" call above will open a window.
+        if (self.as(gio.Application).getIsRemote() != 0) {
+            log.debug(
+                "application is remote, exiting run loop after activation",
+                .{},
+            );
+            return;
+        }
+
+        log.debug("entering runloop", .{});
+        defer log.debug("exiting runloop", .{});
+        priv.running = true;
+        while (priv.running) {
+            _ = glib.MainContext.iteration(ctx, 1);
+
+            // Tick the core Ghostty terminal app
+            try priv.core_app.tick(rt_app);
+
+            // Check if we must quit based on the current state.
+            const must_quit = q: {
+                // If we are configured to always stay running, don't quit.
+                if (!config.@"quit-after-last-window-closed") break :q false;
+
+                // If the quit timer has expired, quit.
+                // if (self.quit_timer == .expired) break :q true;
+
+                // There's no quit timer running, or it hasn't expired, don't quit.
+                break :q false;
+            };
+
+            if (must_quit) {
+                //self.quit();
+                priv.running = false;
+            }
+        }
+    }
+
+    pub fn as(app: *Self, comptime T: type) *T {
+        return gobject.ext.as(T, app);
+    }
+
+    pub fn unref(self: *Self) void {
+        gobject.Object.unref(self.as(gobject.Object));
+    }
```

Comment:
nit: doesn't add much, and if I saw them in code as a new contributor who was familiar with GObject, I'd have to see what they do, versus everyone knows what `g_object_unref()` does.

`as` maybe adds something.

Anyway, it's a nit, so do what you want!

Reply:
Actually on further thought, its kind of necessary to add `unref` because otherwise you're going to be doing `gobject.Object.unref(self.as(gobject.Object));` everywhere which is really terrible. You can't mask the pointer types being different except by calling `as` so... yeah. I think we have to keep it to make Zig palatable.